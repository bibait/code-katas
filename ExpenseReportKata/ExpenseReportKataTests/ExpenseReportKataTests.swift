import Testing
import Foundation
import ApprovalTests_Swift
@testable import ExpenseReportKata

struct ExpenseReportKataTests {

    @Test
    func testPrintReport() throws {
        let expenses: [Expenses] = [
            Expenses(expenseList: [
                Breakfast(amount: 0),
                Dinner(amount: 0),
                Lunch(amount: 0),
                CarRental(amount: 0)
            ]),
            Expenses(expenseList: [
                Breakfast(amount: 1000),
                Dinner(amount: 5000),
                Lunch(amount: 2000),
                CarRental(amount: 999)
            ]),
            Expenses(expenseList: [
                Breakfast(amount: 1001),
                Dinner(amount: 5001),
                Lunch(amount: 2001),
                CarRental(amount: 2001)
            ])
        ]

        let printer = SpyMessagePrinter()
        let sut = ExpenseReport(getDate: { date() }, messagePrinter: printer)
        
        try CombinationApprovals.verifyAllCombinations({ expenses in
            sut.printReport(expenses: expenses)
            return printer.messages
        }, expenses)
    }
    
    // MARK: - Helpers
    
    private class SpyMessagePrinter: MessagePrinter {
        var messages: [String] = []
        
        func printMessage(_ message: String) {
            messages.append(message)
        }
    }
    
    private func date() -> Date {
        let calendar = Calendar.current
        let components = DateComponents(year: 2025, month: 6, day: 15, hour: 12)
        return calendar.date(from: components)!
    }

}
