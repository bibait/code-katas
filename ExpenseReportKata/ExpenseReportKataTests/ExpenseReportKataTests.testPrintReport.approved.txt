[ExpenseReportKata.Expenses(expenseList: [ExpenseReportKata.Breakfast, ExpenseReportKata.Dinner, ExpenseReportKata.Lunch, ExpenseReportKata.CarRental])] -> ["Expense Report 2025-06-15 10:00:00 +0000", "Breakfast\t0\t ", "Dinner\t0\t ", "Lunch\t0\t ", "Car Rental\t0\t ", "Meal Expenses: 0", "Total Expenses: 0"]
[ExpenseReportKata.Expenses(expenseList: [ExpenseReportKata.Breakfast, ExpenseReportKata.Dinner, ExpenseReportKata.Lunch, ExpenseReportKata.CarRental])] -> ["Expense Report 2025-06-15 10:00:00 +0000", "Breakfast\t0\t ", "Dinner\t0\t ", "Lunch\t0\t ", "Car Rental\t0\t ", "Meal Expenses: 0", "Total Expenses: 0", "Expense Report 2025-06-15 10:00:00 +0000", "Breakfast\t1000\t ", "Dinner\t5000\t ", "Lunch\t2000\t ", "Car Rental\t999\t ", "Meal Expenses: 8000", "Total Expenses: 8999"]
[ExpenseReportKata.Expenses(expenseList: [ExpenseReportKata.Breakfast, ExpenseReportKata.Dinner, ExpenseReportKata.Lunch, ExpenseReportKata.CarRental])] -> ["Expense Report 2025-06-15 10:00:00 +0000", "Breakfast\t0\t ", "Dinner\t0\t ", "Lunch\t0\t ", "Car Rental\t0\t ", "Meal Expenses: 0", "Total Expenses: 0", "Expense Report 2025-06-15 10:00:00 +0000", "Breakfast\t1000\t ", "Dinner\t5000\t ", "Lunch\t2000\t ", "Car Rental\t999\t ", "Meal Expenses: 8000", "Total Expenses: 8999", "Expense Report 2025-06-15 10:00:00 +0000", "Breakfast\t1001\tX", "Dinner\t5001\tX", "Lunch\t2001\tX", "Car Rental\t2001\t ", "Meal Expenses: 8003", "Total Expenses: 10004"]
